tower/src/util/mod.rs

```diff
@@ -938,3 +940,12 @@ pub trait ServiceExt<Request>: tower_service::Service<Request> {
 }
 
 impl<T: ?Sized, Request> ServiceExt<Request> for T where T: tower_service::Service<Request> {}
+
+/// Convert an `Option<Layer>` into a `Layer`
+pub fn option_layer<L>(layer: Option<L>) -> Either<L, Identity> {
```

Comment:
An alternative to having a special wrapper function would be  to just add an `impl<L: Layer> for Option<L>`. That might have slightly nicer API ergonomics, since _all_ `Option<Layer>`s just _are_ layers, rather than having to wrap them in `option_layer` calls.

But, the downside to that is that this would have to be implemented in the `tower_layer` crate, rather than in `tower`. This means that we couldn't use `Either`, which is defined in `tower`...

Reply:
I actually like that `option_layer` is explicit about whats happening. When looking at large layer stack it might be obvious which layers are optional and which aren't. Is `TimeoutLayer` applied

Reply:
I actually like that `option_layer` is explicit about whats happening. When looking at large layer stack being built with `ServiceBuilder` it might be obvious which layers are optional and which aren't. Is this `TimeoutLayer` applied all the time or only sometimes? Would have to check the type.

That combined with the annoyances of having to copy+paste `Either` (or depend on the [either](https://crates.io/crates/either) crate) makes me think this approach is fine.

Reply:
@davidpdrsn Yeah, I think that makes sense. I'm also in favor of avoiding `tower-service` changes when they're not absolutely necessary. We can always add an impl in `tower-service` later if we end up deciding that _is_ a good idea...

Reply:
for API consistency's sake, WDYT about adding an `option_service` (or something) that takes an `Option<T: Service>` and returns an [`Optional<T>`](https://docs.rs/tower/0.4.4/tower/util/struct.Optional.html)? That way, we'd have `option_layer` and `option_service` functions. 

I'd add that in a separate PR, though.

Reply:
Yeah that sounds good.